## HW 4
## Following are skeleton code for hw4, no numpy version

In [ ]:
"""
Feel free to write other helper functions
And change the parameters/return values of the function
To make them work better
For the first part, you can't use numpy or similar ml libraries but can use utilities such as math
You can also write a class about neural network and put related functions inside it
for loading data still feel free to use numpy
To make things easier, your data can be in numpy form, but 
in the first part you should not use numpy functionality such as matmul or dot or related matrix/vectorization methods
in the second part feel free to use anything in numpy

EXTREMELY IMPORTANT THINGS:
1) do -1 on y to get the correct labels
2) the first value in the weights theta_1 and theta_2 files is the bias

for the skeleton code you don't have to follow it, they are here only for your reference
but if you have no idea where to start you can use them to guide your way
"""

import numpy as np
import time 
import math
# read in data and weights
# since label data is in fact 1-10, we will change them into 0 to 9
X = np.genfromtxt('ps5_data.csv',delimiter=',')
y = np.genfromtxt('ps5_data-labels.csv',delimiter=',')
y = y-1
W1 = np.genfromtxt('ps5_theta1.csv',delimiter=',')
W2 = np.genfromtxt('ps5_theta2.csv',delimiter=',')
print(X.shape, y.shape, W1.shape, W2.shape)

## 4.2 softmax and sigmoid activation
def softmax(yhat):
    # 4.2

    return result
def sigmoid(z):
    
    return 

def toCategorical(y):
    # convert to one hot

    return y_one_hot

class NeuralNetwork:
    def __init__(self):
        self.W1 = None #W1 should be 401 * 25
        self.W2 = None #W2 should be 26 * 10
    def setWeight(self, W1, W2):
        
    def neuronActivation(self,Aprev,W,with_sigmoid=True):
        # 4.1
        #this will take activation of previous layer and the weight of previous layer to this layer
        #to produce a single activation value

        return a
    
    def final_layer(self, activations):
        # 4.3 activation of the final layer

        return probabilities
    
    def forward(self,X):
        # 4.4 
        #non-vectorized. Given 1 example, return a prediction

        return probabilities
    def classify(self,X): 
        # 4.5, should use 4.4 forward function
        # will classify an image and return a number

        return int(predicted_label)
    def get_error_rate(self, X, y):
        # 4.6 will give error rate on given X, y
        # you can use python's time
        start_time = time.time()

        return error_rate, end_time-start_time
    def giveCost(self, X, y):
        # 4.7
        # gives the value of the cross entropy cost function

        return J

(5000, 400) (5000,) (25, 401) (10, 26)


## Run the program and report error rate, time used for classify all X, and cost

In [17]:
nn = NeuralNetwork()
nn.setWeight(W1.transpose(), W2.transpose())
error_rate, time_used = nn.get_error_rate(X, y)
cost = nn.giveCost(X, y)
print(error_rate, cost, time_used)

0.024800000000000044 0.15284346245189467 13.00907301902771


Error rate, cost and time used should be similar to:

0.0248 0.1528 12.4

## Following are skeleton code for hw4, numpy vectorized version

In [18]:
import numpy as np
import time 
# read in data and weights
# since label data is in fact 1-10, we will change them into 0 to 9

X = np.genfromtxt('ps5_data.csv',delimiter=',')
y = np.genfromtxt('ps5_data-labels.csv',delimiter=',')
y = y-1
W1 = np.genfromtxt('ps5_theta1.csv',delimiter=',')
W2 = np.genfromtxt('ps5_theta2.csv',delimiter=',')
print(X.shape, y.shape, W1.shape, W2.shape)

(5000, 400) (5000,) (25, 401) (10, 26)


In [19]:
## 4.2 softmax and sigmoid activation
def softmax(X, dim):
    return 
def sigmoid(X):
    return 
def toCategorical(y):
    # convert to one hot
    
    return y_one_hot

In [20]:
class NeuralNetwork:
    def __init__(self):
        self.W1 = None #W1 should be 401 * 25
        self.W2 = None #W2 should be 26 * 10
    def setWeight(self, W1, W2):

    def AddBias(self, X):
        #will add the bias term to the input matrix

        return Xnew
    def hidden_layer_activation(self, X, W):
        ## 4.1
        ## use sigmoid activation for hidden layers

        return 
    def final_layer_activation(self, X, W):
        ## 4.3
        ## only applies to final layer
        ## note you only use softmax, don't use sigmoid here

        return 
    def forward(self, X):
        ## 4.4, should use 4.1, 4.3 functions

        return probabilities 
    def classify(self,X): 
        # 4.5, should use 4.4 forward function
        # will classify an image and return a number

        return 
    def get_error_rate(self, X, y):
        # 4.6 will give error rate on given X, y
        start_time = time.time()

        return error_rate, end_time-start_time
    def giveCost(self, X, y):
        # 4.7
        # gives the value of the cross entropy cost function

        return J

## Run the program and report error rate, time used for classify all X, and cost

In [22]:
nn = NeuralNetwork()
nn.setWeight(W1.transpose(), W2.transpose())
error_rate, time_used = nn.get_error_rate(X, y)
cost = nn.giveCost(X, y)
print(error_rate, cost, time_used)

0.024800000000000044 0.152843462451895 0.020461082458496094


Error rate, cost and time used should be similar to:

0.0248 0.1528 0.03

## Is vectorization making things faster?